bolt-modules/boltlib/lib/puppet/functions/apply_prep.rb

```diff
@@ -8,155 +8,206 @@
 # installed using either the configured plugin or the `task` plugin with the
 # `puppet_agent::install` task.
 #
-# Agent installation will be skipped if the target includes the `puppet-agent` feature, either as a
-# property of its transport (PCP) or by explicitly setting it as a feature in Bolt's inventory.
+# Agent installation will be skipped if the target includes the `puppet-agent`
+# feature, either as a property of its transport (PCP) or by explicitly setting
+# it as a feature in Bolt's inventory.
 #
 # > **Note:** Not available in apply block
 Puppet::Functions.create_function(:apply_prep) do
   # @param targets A pattern or array of patterns identifying a set of targets.
   # @param options Options hash.
   # @option options [Array] _required_modules An array of modules to sync to the target.
   # @option options [String] _run_as User to run as using privilege escalation.
-  # @return [nil]
+  # @return [Bolt::ResultSet]
   # @example Prepare targets by name.
   #   apply_prep('target1,target2')
   dispatch :apply_prep do
     param 'Boltlib::TargetSpec', :targets
     optional_param 'Hash[String, Data]', :options
   end
 
-  def script_compiler
-    @script_compiler ||= Puppet::Pal::ScriptCompiler.new(closure_scope.compiler)
+  def apply_prep(target_spec, options = {})
+    unless Puppet[:tasks]
+      raise Puppet::ParseErrorWithIssue
+        .from_issue_and_stack(Bolt::PAL::Issues::PLAN_OPERATION_NOT_SUPPORTED_WHEN_COMPILING, action: 'apply_prep')
+    end
+
+    options = options.slice(*%w[_catch_errors _required_modules _run_as])
+    targets = inventory.get_targets(target_spec)
+
+    executor.report_function_call(self.class.name)
+
+    executor.log_action('install puppet and gather facts', targets) do
+      executor.without_default_logging do
+        install_results = install_agents(targets, options)
+        facts_results   = get_facts(install_results.ok_set.targets, options)
+
+        Bolt::ResultSet.new(install_results.error_set.results + facts_results.results)
+      end
+    end
+  end
+
+  def applicator
+    @applicator ||= Puppet.lookup(:apply_executor)
+  end
+
+  def executor
+    @executor ||= Puppet.lookup(:bolt_executor)
   end
 
   def inventory
     @inventory ||= Puppet.lookup(:bolt_inventory)
   end
 
+  def script_compiler
+    @script_compiler ||= Puppet::Pal::ScriptCompiler.new(closure_scope.compiler)
+  end
+
+  # Retrieves a task by name. This method is called by the puppet_library hook.
+  #
   def get_task(name, params = {})
```

Comment:
https://github.com/puppetlabs/bolt/commit/f4db81f56d6cbee8aa3b04554f886de6d6754327 I don't remember *why* we initially had to load the task here instead of using `pal.get_task` besides for nicer error messages, but I think it'd make more sense to:
* Call `pal.get_task` in the puppet_library plugin
* Call `tasksig.runnable_with?(params)` there as well

The PluginExecution error already includes the information that this is running as a plugin, so better error messaging doesn't seem necessary, and then we're not duplicating this code with `pal.get_task`.